ignition_ros2_control/src/ignition_ros2_control_plugin.cpp

```diff
@@ -0,0 +1,468 @@
+  this->dataPtr->controller_manager_.reset(
+    new controller_manager::ControllerManager(
+      std::move(resource_manager_),
+      this->dataPtr->executor_,
+      "controller_manager"));
+  this->dataPtr->executor_->add_node(this->dataPtr->controller_manager_);
+
+  if (!this->dataPtr->controller_manager_->has_parameter("update_rate")) {
+    RCLCPP_ERROR_STREAM(
+      this->dataPtr->node->get_logger(),
+      "controller manager doesn't have an update_rate parameter");
+    return;
+  }
+
+  auto cm_update_rate = this->dataPtr->controller_manager_->get_parameter("update_rate").as_int();
+  this->dataPtr->control_period_ = rclcpp::Duration(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::duration<double>(1.0 / static_cast<double>(cm_update_rate))));
+
+  this->dataPtr->entity_ = _entity;
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::PreUpdate(
+  const ignition::gazebo::UpdateInfo & _info,
+  ignition::gazebo::EntityComponentManager & /*_ecm*/)
+{
+  static bool warned{false};
+  if (!warned) {
+    rclcpp::Duration gazebo_period(_info.dt);
+
+    // Check the period against the simulation period
+    if (this->dataPtr->control_period_ < _info.dt) {
+      RCLCPP_ERROR_STREAM(
+        this->dataPtr->node->get_logger(),
+        "Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+          " s) is faster than the gazebo simulation period (" <<
+          gazebo_period.seconds() << " s).");
+    } else if (this->dataPtr->control_period_ > gazebo_period) {
+      RCLCPP_WARN_STREAM(
+        this->dataPtr->node->get_logger(),
+        " Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+          " s) is slower than the gazebo simulation period (" <<
+          gazebo_period.seconds() << " s).");
+    }
+    warned = true;
+  }
+
+  // Always set commands on joints, otherwise at low control frequencies the joints tremble
+  // as they are updated at a fraction of gazebo sim time
+  this->dataPtr->controller_manager_->write();
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::PostUpdate(
+  const ignition::gazebo::UpdateInfo & _info,
+  const ignition::gazebo::EntityComponentManager & /*_ecm*/)
+{
+  // Get the simulation time and period
+  rclcpp::Time sim_time_ros(std::chrono::duration_cast<std::chrono::nanoseconds>(
+      _info.simTime).count(), RCL_ROS_TIME);
+  rclcpp::Duration sim_period = sim_time_ros - this->dataPtr->last_update_sim_time_ros_;
+
+  if (sim_period >= this->dataPtr->control_period_) {
+    this->dataPtr->last_update_sim_time_ros_ = sim_time_ros;
+    auto ign_controller_manager =
+      std::dynamic_pointer_cast<ignition_ros2_control::IgnitionSystemInterface>(
+      this->dataPtr->controller_manager_);
+    this->dataPtr->controller_manager_->read();
+    this->dataPtr->controller_manager_->update();
```

Comment:
It seems to me that the following line is missing. Can this be? Or I am getting something wrong?

```suggestion
    this->dataPtr->controller_manager_->update();
    this->dataPtr->controller_manager_->write();
```

We should execute updates something like `ros2_control_node` is doing. (check [here](https://github.com/ros-controls/ros2_control/blob/a47a0347acd28a47b5095e4b206257e1520918ee/controller_manager/src/ros2_control_node.cpp#L51-L53))